### PR TITLE
Winnow private method candidates instead of assuming any candidate of the right name will apply

### PR DESCRIPTION
**Pull request by oli-obk**
partially reverts https://github.com/rust-lang/rust/pull/60721

My original motivation was just to avoid the `delay_span_bug` (by attempting to thread the `ErrorGuaranteed` through to here). But then I realized that the error message is wrong. It refers to the `Foo<A>::foo` instead of `Foo<B>::foo`. This is almost invisible, because both functions are the same, but on different lines, so `-Zui-testing` makes it so the test is the same no matter which of these two functions is referenced.

But there's a much more obvious bug: If `Foo<B>` does not have a `foo` method at all, but `Foo<A>` has a private `foo` method, then we'll refer to that one. This has now been fixed, and we report a normal `method not found` error.

The way this is done is by creating a list of all possible private functions (just like we create a list of the public functions that can actually be called), and then winnowing it by analyzing where bounds and `Self` types to see if any of the found methods can actually apply (again, just like with the list of public functions).

I wonder if there is room for doing the same thing with unstable functions instead of running all of method resolution twice.

r? @compiler-errors for method resolution stuff

### CODE REVIEW

**Comment by compiler-errors** [comment]
Can you please share what the motivation here is? Don't see a linked issue and there's no inline discussion, and there's also no test -- hard to understand how all these changes fit together without that.

@rustbot author

File: compiler/rustc_hir_typeck/src/method/probe.rs [hunk 1]
```diff
@@ -76,8 +77,12 @@ pub(crate) struct ProbeContext<'a, 'tcx> {
     /// requested name (by edit distance)
     allow_similar_names: bool,
 
+    /// List of potential private candidates. Will be trimmed to ones that
+    /// actually apply and then the result inserted into `private_candidate`
+    private_candidates: Vec<Candidate<'tcx>>,
+
     /// Some(candidate) if there is a private candidate
-    private_candidate: Option<(DefKind, DefId)>,
+    private_candidate: Cell<Option<(DefKind, DefId)>>,
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/method/probe.rs:85]: What is the difference between these lol

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/method/probe.rs:85]: Or actually, better question: Why do we need two fields here?

**Comment by oli-obk** [comment]
> there's also no test

🤦 there is a test, I've just been looking at it all day working on this, and the end result was "no change", which I wanted.

Though I just remembered how to write a test that has user visible changes.  Will update docs and tests tomorrow

**Comment by compiler-errors** [comment]
@oli-obk: I don't need a test per se, but eitehr a test or motivation for why you want this change otherwise would be useful. I just don't know how to review this change w/o any information, since it doesn't seem like a cleanup by itself.

File: compiler/rustc_hir_typeck/src/method/probe.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_hir_typeck/src/method/probe.rs:85]: We need the `private_candidate` field to communicate an actually found candidate out to the `pick` method. While we could just eagerly return the `PrivateMatch` error, that would mean we'd miss out on suggesting traits that could be imported to get access to another method of the same name.

Of course we could just eagerly report (not return) an error from method selection and return that no methods were found, which will subsequently return an error containing the list of traits that should be imported. But then method selection is reporting errors, which may also not be desirable.

**Comment by compiler-errors** [comment]
@bors r- https://github.com/rust-lang/rust/pull/125975#issuecomment-2147670275

double check tidy perhaps?

**Comment by oli-obk** [comment]
rebased and adjusted tidy limits

@bors r=compiler-errors

**Comment by fmease** [comment]
bors sleepy @bors r-

**Pull request merged**